lbrynet/core/Wallet.py

```diff
@@ -130,54 +210,213 @@ def create_tables(transaction):
-            (claim_id, name, claim_outpoint['txid'], claim_outpoint['nout']))
-        d.addCallback(
-            lambda _: self.db.runQuery(
-                "delete from claim_ids where claimId=? and name=? and txid=? and n=?",
-                (claim_id, name, claim_outpoint['txid'], UNSET_NOUT)))
-        d.addCallback(
-            lambda r: self.db.runQuery(
-                "insert into claim_ids values (?, ?, ?, ?)",
-                (claim_id, name, claim_outpoint['txid'], claim_outpoint['nout'])))
-        d.addCallback(lambda _: claim_id)
-        return d
+        txid, nout = claim_outpoint['txid'], claim_outpoint['nout']
+        result = yield self.db.runQuery("SELECT COUNT(*) FROM claim_ids WHERE claimId=?",
+                                           (claim_id,))
+        is_update = result[0][0]
+        if is_update:
+            yield self.db.runOperation("UPDATE claim_ids SET name=?, txid=?, n=? WHERE claimId=?",
+                                       (name, txid, nout, claim_id))
+        else:
+            yield self.db.runOperation("INSERT INTO claim_ids VALUES (?, ?, ?, ?)",
+                                       (claim_id, name, txid, nout))
+        defer.returnValue(claim_id)
 
-    def get_claimid_for_tx(self, name, claim_outpoint):
-        d = self.db.runQuery(
-            "select claimId from claim_ids where name=? and txid=? and n=?",
-            (name, claim_outpoint['txid'], claim_outpoint['nout']))
-        d.addCallback(lambda r: r[0][0] if r else None)
-        return d
+    @rerun_if_locked
+    @defer.inlineCallbacks
+    def get_claimid_for_tx(self, claim_outpoint):
+        result = yield self.db.runQuery("SELECT claimId FROM claim_ids "
+                                        "WHERE txid=? AND n=?",
+                                        (claim_outpoint['txid'], claim_outpoint['nout']))
+        response = None
+        if result:
+            response = result[0][0]
+        defer.returnValue(response)
+
+    @rerun_if_locked
+    @defer.inlineCallbacks
+    def get_cached_claim(self, claim_id, check_expire=True):
+        r = yield self.db.runQuery("SELECT * FROM claim_cache WHERE claim_id=?", (claim_id, ))
+        claim_tx_info = yield self.db.runQuery("SELECT name, txid, n FROM claim_ids "
+                                               "WHERE claimId=?", (claim_id, ))
+        response = None
+        if r and claim_tx_info:
+            _, _, seq, claim_address, height, amount, supports, raw, chan_name, valid, ts = r[0]
+            last_modified = int(ts)
+            supports = [] if not supports else json.loads(supports)
+            name, txid, nout = claim_tx_info[0]
+            if check_expire and (time.time() - last_modified) > CLAIM_CACHE_TIME:
+                log.debug("Cached claim expired: %s", claim_id)
+            else:
+                deserialized = ClaimDict.deserialize(raw.decode('hex'))
+                response = {
+                    "height": height,
```

Comment:
is this a duplicate of https://github.com/lbryio/lbry/pull/620/files#diff-eafee006192cca2ffbca5ba91dc83eeeR145 ? something feels wrong about returning this dict here (and above too). shouldn't we return an object? the function is called `get_cached_claim`. we should be returning a claim, not a dict with some arbitrary fields